zerver/lib/actions.py

```diff
@@ -4735,6 +4742,16 @@ def do_change_user_role(
     )
     send_event(user_profile.realm, event, active_user_ids(user_profile.realm_id))
 
+    UserGroupMembership.objects.filter(
+        user_profile=user_profile, user_group=old_system_group
+    ).delete()
+
+    system_group = get_system_user_group_for_user(user_profile)
+    UserGroupMembership.objects.create(user_profile=user_profile, user_group=system_group)
```

Comment:
Do we want this part in `transaction.atomic` block to avoid the groups being in inconsistent state?

Reply:
You can add several `on_commit` hookos; see https://docs.djangoproject.com/en/3.2/topics/db/transactions/#order-of-execution for a brief refrence to that being supported.